hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ListIterator.java

```diff
@@ -0,0 +1,355 @@
+    }
+
+    public void close() {
+      // Nothing to close here
+    }
+  }
+
+  /**
+   * Implement lexicographical sorting of the file status by sorting file status
+   * across multiple lists. Each of these lists are sorted internally.
+   *
+   * This class implements sorted output by implementing a min heap based
+   * iterator where the initial element from each of sorted list is inserted.
+   *
+   * The least entry is removed and the next entry from the same list from
+   * which the entry is removed is added into the list.
+   *
+   * For example
+   * RawDir   - a1, a3, a5, a7
+   * RawFile  - a2, a4, a6, a8
+   *
+   * Min Heap is initially composed of {(a1, RawDir), (a2, RawFile)}
+   * THe least element is removed i.e a1 and then next entry from RawDir
+   * is inserted into minheap resulting in {(a2, RawFile), (a3, RawDir)}
+   *
+   * This process is repeated till both the lists are exhausted.
+   */
+  public static class MinHeapIterator implements ClosableIterator {
+    private final PriorityQueue<HeapEntry> minHeap = new PriorityQueue<>();
+    private final ArrayList<ClosableIterator> iterators = new ArrayList<>();
+
+    MinHeapIterator(OMMetadataManager omMetadataManager, String prefixKey,
+                    BucketLayout bucketLayout, String startKey,
+                    String volumeName, String bucketName) throws IOException {
+
+      this(omMetadataManager, prefixKey, startKey, volumeName,
+          bucketName, omMetadataManager.getDirectoryTable(),
+          omMetadataManager.getKeyTable(bucketLayout));
+    }
+
+    MinHeapIterator(OMMetadataManager omMetadataManager, String prefixKey,
+                    String startKey, String volumeName, String bucketName,
+                    Table... tables) throws IOException {
+      omMetadataManager.getLock().acquireReadLock(BUCKET_LOCK, volumeName,
+          bucketName);
+      try {
+        int iteratorId = 0;
+        for (Table table : tables) {
+          iterators.add(new CacheIter<>(iteratorId, table.getName(),
+                  table.cacheIterator(), startKey, prefixKey));
+          iteratorId++;
+          iterators.add(new RawIter<>(iteratorId,
+                  omMetadataManager.getDirectoryTable(),
+                  prefixKey, startKey));
+          iteratorId++;
+        }
+      } finally {
+        omMetadataManager.getLock().releaseReadLock(BUCKET_LOCK, volumeName,
+            bucketName);
+      }
+
+      // Insert the element from each of the iterator
+      for (Iterator<HeapEntry> iter : iterators) {
+        if (iter.hasNext()) {
+          minHeap.add(iter.next());
+        }
+      }
+
+    }
+
+    public boolean hasNext() {
+      return !minHeap.isEmpty();
+    }
+
+    public HeapEntry next() {
+      HeapEntry heapEntry = minHeap.remove();
+      // remove the least element and
+      // reinsert the next element from the same iterator
+      Iterator<HeapEntry> iter = iterators.get(heapEntry.getEntryInteratorId());
```

Comment:
Correct me if I'm missing something but I think we are gonna return some entries twice.
For example if cacheTable has {e3, e4, e5}  and dbTable has {e1, e2, e3, e4, e5} , minHeap iterator would return {e1, e2, e3, e3, e4, e4, e5, e5}.

Reply:
yeah I guess there was a bug in the previous code as well. This was being used for listing keys & dirs for FSO buckets.

Reply:
I think it is still possible that something will be returned twice in the updated code.

Let's take the previous example, cacheTable has {e3, e4, e5} and dbTable has {e1, e2, e3, e4, e5} and page size is 3. First page would be {e1, e2, e3}. e1 and e2 are from DB and e3 from cache. When we get request for second page, cacheTable is {e4, e5, e6} and dbTable is {e1, e2, e3, e4, e5}. Second page response would be {e3, e4, e5}. e3 form DB and e4 and e5 from either cache or DB.

Reply:
Yes, you are right.